### DISCUSSION TITLE
memory lack

### DISCUSSION THREAD

**Discussion by kenan991** [pybamm-team/PyBaMM#1937] [Q&A]
Hello All
I am using pybamm.lithium_ion.DFN() and implementing  the experiment on my data set, I mean I have data set(3000 values) containing power values and in every iteration, the experiment is dis/charge depending on the sign of the value but after 1000 iterations I am getting a memory lack (to know I have 8 Giga RAM) I tried to solve this problem by getting lose of unwanted previous values which saved in the arrays in the simulation class I mean (all_ts, all_ys, all_models, all_inputs, t_event, y_event) but I still get the same memory lack.
that is what I mean by simulation class
https://pybamm.readthedocs.io/en/latest/source/solvers/solution.html?highlight=solution#pybamm.Solution

please, is there a way to solve my problem?

Best
Kenan

**Comment by valentinsulzer** [comment]
What do you mean by "iteration" here? Cycles? Can you provide a minimum working example?

**Reply by kenan991** [reply]
yeah I mean cycles

this part of my code where I implement the battery using pybamm

class Battery:  # check the power drwan/feeded to the grid

    def __init__(self,  Discharge capacity=0):

        #self.last_solution = None
        self.parameter_values = pybamm.ParameterValues(chemistry=pybamm.parameter_sets.Chen2020)
        self.model = pybamm.lithium_ion.DFN()

        # TODO use these to initi
        self.Discharge capacity = Discharge capacity

    def update(self, power):
        global last_solution
        #time_passed = self.env.now - self.last_update
        if power > 0:
            experiment = pybamm.Experiment([f"charge at {0.1 * power} W for 15 s"])
        elif power < 0:
            experiment = pybamm.Experiment([f"discharge at {-0.1 * power} W for 15 s"])
        else:
            return 0

        sim = pybamm.Simulation(self.model, parameter_values=self.parameter_values, experiment=experiment)
        #sim.solve(starting_solution=self.last_solution)
        sim.solve(starting_solution=last_solution)
        last_solution = sim.solution
        #self.last_solution = sim.solution
        d = last_solution['Discharge capacity [A.h]']
        self.Discharge capacity =d.entries[-1]
        sim.plot(['Current [A]','Terminal voltage [V]','Discharge capacity [A.h]'])

battery=Battery()
for i in range(0, len( delta_energy )):   ##len(production_df)=2880
        grid_energy = battery.update(delta_energy)


so it works well until i=1000 I mean after making 1000 cycle and then the simulation is slowing up and the memory gets full
and could not finish the simulation for the whole data set

**Reply by valentinsulzer** [reply]
Check out some of the examples in https://github.com/pybamm-team/PyBaMM/blob/develop/examples/notebooks/simulating-long-experiments.ipynb, especially `save_at_cycles`
